### ISSUE TITLE
T5 inference in Candle

### ISSUE THREAD

**Issue by shivance** [huggingface/candle#621]
I'm looking forward to try candle out for inference.

In particular I want to run inference of a `t5-base` model on my mac. What's recommended approach?

Would I need to rewrite entire T5 model in rust using Candle's APIs?

**Comment by LaurentMazare** [comment]
Yes the idea would be to rewrite your t5 model using the candle api - at some point we may supporting for format such as onnx but that's not handled yet.
If you were to start with this, note that we have some preliminary version of a T5 model as part of the `musicgen` example (which is not working yet), so you can probably start by using the following code: https://github.com/huggingface/candle/blob/main/candle-examples/examples/musicgen/t5_model.rs

**Comment by shivance** [comment]
Thanks for the reference @LaurentMazare . It helps. 
Question 2.: The candle framework is new, and I really love this initiative. Is there anywhere I can contribute? If yes, then where (do we have a roadmap?)

**Comment by LaurentMazare** [comment]
We don't really have an explicit roadmap at the moment as things are very much in flux and moving around.
I think a great way to get involved would be to use candle to build the models that you care about, e.g. `t5-base`, in a way that supports text inference would be great! We provide some examples with candle but having "external" examples would provide more diverse use cases and help tune the experience of writing new models (and we likely will need more educational resources on this). Also by writing such models, we'll learn about the missing bits that should be added to candle, the current limitations, the bits that could go to some shared library, etc. And ideally we would have some form of candle model zoo/hub to make it easy to discover all the models written with candle.

**Comment by LaurentMazare** [comment]
t5 inference should now be available thanks to #864 (kudos to @jbochi ).

**Comment by LaurentMazare** [comment]
Closing was hopefully it's all good now but feel free to reopen or create new entries if you run into any issue.

**Comment by shivance** [comment]
Awesome @LaurentMazare !! Is there any benchmark of candle T5 Model available?

**Comment by jbochi** [comment]
I did some local tests on my Apple M2 Pro and the implementation from transformers is a bit faster. Using `t5-base`, a prompt of 42 tokens, and 25 decoded tokens, I got ~18 tokens/s with candle and 22 tokens/s with transformers.

**Comment by LaurentMazare** [comment]
Things to keep in mind when benchmarking on macos: candle doesn't support metal yet, depending on your pytorch setup it may well be used on the other side; it's recommended to run with `--feature accelerate`, this could result in some large speedups. I haven't done any specific performance tuning for t5 but will start tweaking it a bit (and at least adding better tracing so that it's easier to understand where the time is spent).

**Comment by jbochi** [comment]
Thanks @LaurentMazare !

With `--features accelerate`, I got 34 tokens/s, so already faster than the `transformers` implementation.

If you can speed it up any further, that would be great!

**Comment by okpatil4u** [comment]
Here is another approach. Set num_rayon_threads=1, then run it on every
core manually. Your final token throughput
 should be multiple of number of cores as long as you are using only
performance cores on M1/M2.

**Comment by LaurentMazare** [comment]
I've pushed some tracing changes to t5, so if you run the examples with `--tracing` you can a trace that can be viewed as follows in chrome (beware that running with tracing has some small impact on performance).

What is interesting here is that ~85% is spent in linear and lm-head which are matrix multiplications optimized by accelerate so these are likely tricky to optimize. The kv-cache concatenation takes ~8% of the time. Maybe this could be improved but even if it's the case thi will not end up with a very large speedup I'm afraid.

![image](https://github.com/huggingface/candle/assets/1041292/66bb0216-dc9f-4a13-9bb9-9e4a6da426cd)

**Comment by jbochi** [comment]
Very interesting analysis. Would it be possible to avoid the full matrix
multiplication by doing ANN search with a dot product index? At least for
greedy decoding we don't need exact logits for every token.

**Comment by LaurentMazare** [comment]
That's an interesting idea though I'm not sure whether it would be worth the additional complexity. If you want to take a stab at it, you can probably use a `CustomOp` for this, see this [example](https://github.com/huggingface/candle/blob/ab1d40ea97b387f0dd05f77db37c840a4d624a08/candle-examples/examples/custom-ops/main.rs#L33) (in a nutshell, custom ops make it easy to write your own cpu/cuda kernels, typically to be used for ops where performance matters a lot).

**Comment by LaurentMazare** [comment]
Also I've just added a quantized version of the t5 example. The goal is more to show case how easy it is with candle to quantize arbitrary models (hopefully much easier than with llama.cpp).
However on the current variant (t5-small quantized with Q6K) it's slower on my mac than the non-quantized version when using accelerate. On x86 it seems much faster though. Probably worth trying on larger t5 variants as it's likely to make more of a difference there.

You can give this a spin with the following and the [readme](https://github.com/huggingface/candle/tree/main/candle-examples/examples/quantized-t5) has some details on how to generate the quantized weight files if one wants to try different models / quantization types.
```bash
cargo run --example quantized-t5 --profile=release-with-debug --  --prompt "translate to German: A beautiful candle that casts a large halo of yellow light."
```

Also adding @LLukas22 who mentioned some interest in quantizing existing models.

**Comment by okpatil4u** [comment]
Thanks Laurent. Does flan-t5-xxl have a different architecture altogether?

We are finding some issues in generation.

**Comment by LaurentMazare** [comment]
Yes it uses multiple safetensor files as input so as not to get above some 10GB limit. The current code assumes a single file so it's unlikely to work without some tweaking. Are these the issues you are seeing? (flat-t5-large should work fine as it's only 2.92GB)

**Comment by okpatil4u** [comment]
Apparently the logits_processor.sample function is always predicting token
number number 16814, resulting in decoded value Fenster.  Even our own
softmax and argmax functions are giving the same results.

Can you give it a try at your end ?

**Comment by LaurentMazare** [comment]
That's odd, I just gave flan-t5-xxl a try and it worked all well. Do you have a way for me to reproduce the issue?
```bash
cargo run --example t5 --profile=release-with-debug --  --prompt "translate to German: A beautiful candle that casts a large halo of yellow light." --decod
e --model-id google/flan-t5-xxl
 Ein wundervolles Feuerwerk, der einen großen Handel grauer Licht hinterlässt.
```
The quantisation bits have been merged, you will find some links to the readme/example above in this thread.

**Comment by okpatil4u** [comment]
Yes, this works for me. This gives an ability to debug what went wrong in
our code. Thanks for your help !
